core/lib/mini_merkle_tree/src/tests.rs

```diff
@@ -205,15 +262,127 @@ fn merkle_proofs_are_valid_in_very_small_trees() {
+            .parse()
+            .unwrap();
+
+    let expected_path = [
+        "0x39f19437665159060317aab8b417352df18779f50b68a6bf6bc9c94dff8c98ca",
+        "0xc3d03eebfd83049991ea3d3e358b6712e7aa2e2e63dc2d4b438987cec28ac8d0",
+        "0xe3697c7f33c31a9b0f0aeb8542287d0d21e8c4cf82163d0c44c7a98aa11aa111",
+        "0x199cc5812543ddceeddd0fc82807646a4899444240db2c0d2f20c3cceb5f51fa",
+        "0x6edd774c0492cb4c825e4684330fd1c3259866606d47241ebf2a29af0190b5b1",
+        "0x29694afc5d76ad6ee48e9382b1cf724c503c5742aa905700e290845c56d1b488",
+    ]
+    .map(|s| s.parse::<H256>().unwrap());
+
+    for i in 0..50 {
+        let (root_hash, path) = tree.merkle_root_and_path(49 - i);
+        assert_eq!(root_hash, expected_root_hash);
+        assert_eq!(path, expected_path);
+        tree.trim_start(1);
+    }
+
+    let mut tree = MiniMerkleTree::new(leaves, None);
+    for i in 0..10 {
+        let (root_hash, path) = tree.merkle_root_and_path(49 - i * 5);
+        assert_eq!(root_hash, expected_root_hash);
+        assert_eq!(path, expected_path);
+        tree.trim_start(5);
+    }
+}
+
+#[test]
+#[allow(clippy::cast_possible_truncation)] // truncation is intentional
+fn pushing_new_leaves() {
+    let mut tree = MiniMerkleTree::new(iter::empty(), None);
+
+    let expected_roots = [
+        "0x6f7a80e6ee852bd309ee9153c6157535092aa706f5c6e51ff199a4be012be1fd",
+        "0xda895440272a4c4a0b950753c77fd08db0ce57e21c98b75d154c341cbe5f31ac",
+        "0x74e62d47c142e2a5b0f2c71ea0f8bcca8d767f0edf7ec7b9134371f5bfef7b8a",
+        "0xe44bb0f3915370e8f432de0830c52d5dc7bbf1a46a21cccb462cefaf3f4cce4d",
+        "0x88443c3b1b9206955625b5722c06bca3207d39f6044780af885d5f09f6e615a1",
+    ]
+    .map(|s| s.parse::<H256>().unwrap());
+
+    for (i, expected_root) in expected_roots.iter().enumerate() {
+        let number = i as u8 + 1;
+        tree.push([number; 88]);
+        tree.push([number; 88]);
+        tree.push([number; 88]);
+
+        let (root, start_path, end_path) = tree.merkle_root_and_paths_for_range(1);
+        assert_eq!(root, *expected_root);
+        assert_eq!(start_path.len(), end_path.len());
+
+        tree.trim_start(2);
+
+        let (root, start_path, end_path) = tree.merkle_root_and_paths_for_range(1);
+        assert_eq!(root, *expected_root);
+        assert_eq!(start_path.len(), end_path.len());
+    }
+}
+
+#[test]
+fn trim_all_and_grow() {
+    let mut tree = MiniMerkleTree::new(iter::repeat([1; 88]).take(4), None);
+    tree.trim_start(4);
+    tree.push([1; 88]);
+    let expected_root = "0xfa4c924185122254742622b10b68df8de89d33f685ee579f37a50c552b0d245d"
+        .parse()
+        .unwrap();
+    assert_eq!(tree.merkle_root(), expected_root);
+}
+
+#[test]
+fn trim_all_and_check_root() {
+    let mut tree = MiniMerkleTree::new(iter::repeat([1; 88]).take(4), None);
+    let root = tree.merkle_root();
+    tree.trim_start(4);
+    assert_eq!(tree.merkle_root(), root);
```

Comment:
Nit: Similar to the test above, you can test this more thoroughly, e.g. by testing trees with varying number of leaves:

```rust
for len in 1..=100 {
    let mut tree = MiniMerkleTree::new(iter::repeat([1; 88]).take(len), None);
    let root = tree.merkle_root();
    tree.trim_start(len);
    assert_eq!(tree.merkle_root(), root);
}
```